### PR TITLE
Introduce warmup_ratio training argument

### LINKED ISSUE #6673: New training arg: warmup_ratio
# 🚀 Feature request

When training or fine-tuning a transformer model, people usually warmup for 10% training steps. For now, transformers only provide the parameter of warmup_steps. A warmup_ratio parameter can be helpful, it means warmup for some percentage of total training steps.

## Motivation

To use the parameter of warmup_steps, people need to know the total training steps. When people use the training epoch parameter instead of max_steps, it is hard to know the total training steps. A warmup_ratio parameter get rid of people knowing total training steps. Another reason for using warmup_ratio parameter is it can help people write less hard code. People have different total training steps for different dataset, but people usually set warmup_ratio as 10% as default.

Original Usage may like this:
`
python run_ner.py --data_dir some_data_dir \
--model_name_or_path some_model \
--output_dir some_output_dir \
--max_seq_length 512 \
--num_train_epochs 10 \
--warmup_steps 35 \
--per_device_train_batch_size 8 \
--do_train \
`

New usage may like this:
`
python run_ner.py --data_dir some_data_dir \
--model_name_or_path some_model \
--output_dir some_output_dir \
--max_seq_length 512 \
--num_train_epochs 10 \
--warmup_ratio 0.1 \
--per_device_train_batch_size 8 \
--do_train \
`

Also, we can merge warmup_step and warmup_ratio into one parameter. If user input a number 0 <= x < 1, it will be considered as warmup_ratio. If user input an interger, it will be considered as warmup_step.

## Your contribution

I can submit a PR to complete this feature. If similar feature is alreadly in this repo, please just close this issue.

### PR DESCRIPTION
**Pull request by tanmay17061**
Introduce warmup_ratio training argument in both
TrainingArguments and TFTrainingArguments classes (#6673)

# What does this PR do?

This PR will add a new argument `warmup_ratio` to both `TrainingArguments` and `TFTrainingArguments` classes. This can be used to specify the ratio of total training steps for which linear warmup will happen.  
This is especially convenient when the user wants to play around with the `num_train_epochs` or `max_steps` arguments while keeping the ratio of warmup steps a constant.
Fixes #6673

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [Link to the issue raised](https://github.com/huggingface/transformers/issues/6673).
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

Since modifications in trainer: @sgugger

### CODE REVIEW

**Comment by tanmay17061** [comment]
As per current implementation, any non-zero value given for `warmup_steps` will override any effects of `warmup_ratio`. It made sense for me to give higher precedence to `warmup_steps` as it seems to be the more inconvenient argument of the 2 to provide from user perspective. Please let me know if this default behaviour is to be changed.

PS: this is my first PR, so feel free to correct me. I will be happy to accommodate 😄

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -811,6 +816,11 @@ def train(
             num_train_epochs = 1
             num_update_steps_per_epoch = max_steps
 
+        if self.args.warmup_ratio > 0 and self.args.warmup_steps == 0:
+            if self.args.warmup_ratio > 1:
+                raise ValueError("warmup_ratio must lie in range [0,1]")
+            self.args.warmup_steps = math.ceil(max_steps * self.args.warmup_ratio)
```
Comment (sgugger) [inline src/transformers/trainer.py:822]: We don't change the `args` passed by the user -> if they are re-used for another `Trainer` there will be some surprise on the user side. This test should be made when the optimizer is created, to pass the right value for `warmup_steps`.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -811,6 +816,11 @@ def train(
             num_train_epochs = 1
             num_update_steps_per_epoch = max_steps
 
+        if self.args.warmup_ratio > 0 and self.args.warmup_steps == 0:
+            if self.args.warmup_ratio > 1:
+                raise ValueError("warmup_ratio must lie in range [0,1]")
```
Comment (sgugger) [inline src/transformers/trainer.py:821]: This part should go in an `elif` at init: it's best to detect invalid arguments at the very beginning.

File: src/transformers/trainer_tf.py [hunk 3]
```diff
@@ -486,6 +491,11 @@ def train(self) -> None:
         # Since ``self.args.num_train_epochs`` can be `float`, we make ``epochs`` be a `float` always.
         epochs = float(epochs)
 
+        if self.args.warmup_ratio > 0 and self.args.warmup_steps == 0:
+            if self.args.warmup_ratio > 1:
+                raise ValueError("warmup_ratio must lie in range [0,1]")
+            self.args.warmup_steps = math.ceil(t_total * self.args.warmup_ratio)
```
Comment (sgugger) [inline src/transformers/trainer_tf.py:497]: Same comments as above.

**Review by sgugger — APPROVED** [review]
Thanks for your PR! This will make a useful addition!
I left a few comments on how to reorganize a bit differently the code. The main issue is that we don't want to change the values in `self.args`. Let me know if you run in any issue when addressing them.

**Comment by tanmay17061** [comment]
Thanks for the review! I've incorporated comments.  
Do let me know if anything else needs to be addressed.

File: src/transformers/trainer_tf.py [hunk 4]
```diff
@@ -221,7 +228,9 @@ def create_optimizer_and_scheduler(self, num_training_steps: int):
             self.optimizer, self.lr_scheduler = create_optimizer(
                 self.args.learning_rate,
                 num_training_steps,
-                self.args.warmup_steps,
+                self.args.warmup_steps
+                if self.args.warmup_steps > 0
+                else math.ceil(num_training_steps * self.args.warmup_ratio),
```
Comment (sgugger) [inline src/transformers/trainer_tf.py:233]: Small nit: this is not super readable so it would be great to define before this function
```
wamup_steps = self.args.warmup_steps if self.args.warmup_steps > 0 else math.ceil(num_training_steps * self.args.warmup_ratio)
```
(which will be restyled since it's a bit long so you will need to run `make style`) and then use `warmup_steps` here.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -609,7 +616,9 @@ def create_optimizer_and_scheduler(self, num_training_steps: int):
             self.lr_scheduler = get_scheduler(
                 self.args.lr_scheduler_type,
                 self.optimizer,
-                num_warmup_steps=self.args.warmup_steps,
+                num_warmup_steps=self.args.warmup_steps
+                if self.args.warmup_steps > 0
+                else math.ceil(num_training_steps * self.args.warmup_ratio),
```
Comment (sgugger) [inline src/transformers/trainer.py:621]: Small nit: this is not super readable so it would be great to define before this function
```
wamup_steps = self.args.warmup_steps if self.args.warmup_steps > 0 else math.ceil(num_training_steps * self.args.warmup_ratio)
```
(which will be restyled since it's a bit long so you will need to run make style) and then use `warmup_steps` here.

**Review by sgugger — APPROVED** [review]
Great! Some last comments on the style (the test is a bit hard to read in the current version) but this is exactly what I had in mind, thanks for adjusting!

**Comment by tanmay17061** [comment]
Thanks for your comments!  
Agreed, the code looks much more readable now.  
Do let me know if there can be any more improvement.  

Thanks.

**Comment by sgugger** [comment]
It would indeed be better in `TrainingArguments.__post_init__`: the rational for that is that when instantiating an object with wrong values, we want the error to be raised as soon as possible and as close as possible to the source for easy debugging.

In this case, the problem should appear at the line that parses the `TrainingArguments` or when they are created.

**Comment by tanmay17061** [comment]
Thanks! Taken care of it.

**Comment by sgugger** [comment]
Thanks for adding this functionality!
